### PR TITLE
Template Part: Improve insertion flow.

### PR DESCRIPTION
**Pull request by epiqueras**
## Description

This PR improves the template part flow by:

- Renaming "Template Part" to "Section" in the UI.
- Implementing a new placeholder design.
- Providing an input to rename existing template parts.
- Improving the empty state to match the group block.

We should follow this up with an exploration of moving the renaming input to the toolbar.

@MichaelArestad @shaunandrews Do you think this flow will change frequently enough that it's worth disabling E2E tests for it, for now, to ease iteration?

## How to test this?

Play around with inserting "Section" blocks in the FSE experiment.

## Screenshots

gif

## Checklist:

- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-library/src/template-part/edit/inner-blocks.js
Comment (Addison-Stavlo) [inline packages/block-library/src/template-part/edit/inner-blocks.js:24]: Should this function be defined as opposed to anonymous?

I think anonymous functions passed into component props cause unnecessary re-renders as `() => {}  !==  () => {}` and the component always views that prop as having changed?

File: packages/block-library/src/template-part/edit/name-panel.js
Comment (Addison-Stavlo) [inline packages/block-library/src/template-part/edit/name-panel.js:27]: Same question here as above.  While passing an anonymous function into an element's `onChange` seems perfectly fine, with the component wouldn't this potentially cause unnecessary re-renders?

**Review by Addison-Stavlo — COMMENTED** [review]
This is great so far!  I really like it.

A couple questions about the dropdown:
* Should the Dropdown / preview panel be wider?  It seems very thin, but it does behave well and it doesn't necessarily look bad.
* There are some circumstances where the dropdown has 2 scroll bars.  Maybe the height of the preview container should be contained within the dropdown?  That is, so the outer scrollbar wouldn't be triggered, the 'search' input would always be at the top of the dropdown, and the 'scroll' would only happen through the preview window itself?

double-scroll-bar


I really like the feel of how the new template part is inserted.  Both having the renaming input and getting rid of the default paragraph seem like a big step up.

This name panel also helps with #22064, making it more apparent that you are editing a template part.  (I suppose I need to get used to calling them 'Sections' 😆  )

<hr/>

> Do you think this flow will change frequently enough that it's worth disabling E2E tests for it, for now, to ease iteration?

After this gets updated for whatever immediate concerns design has, id think it would be worthwhile to update the tests again.  If you want to skip them for now, Id be happy to help update the tests as a follow up (if it makes sense and things seem semi-stable for the time being).  

We should probably try to keep the safety net in place if there is going to be time between iterations.  Id also assume a fair amount of the main concepts of this design will remain in place for a while, in which case 🤞  we are doing 'minor' updates to the tests going forward after this as opposed to re-writing them.

**Review by Addison-Stavlo — COMMENTED** [review]
Also noting nested template parts will trigger the input for their parents.  Im not sure if this is good or bad?  Probably good in that it improves the visibility that the selected template part is part of another?

Screen Shot 2020-06-19 at 12 48 49 PM

**Comment by MichaelArestad** [comment]
This is already a great improvement. I think we'll need to stick to using the term "Template Part" for this PR. Let me know when you're ready for a review.

> Do you think this flow will change frequently enough that it's worth disabling E2E tests for it, for now, to ease iteration?

@epiqueras I don't think functionality will change drastically, but perhaps some of the stuff around reusable block UI or the inserter could change. I trust your judgement here.

File: packages/block-library/src/template-part/edit/inner-blocks.js
Comment (epiqueras) [reply packages/block-library/src/template-part/edit/inner-blocks.js]: Only if the component is "Pure", `InnerBlocks` isn't last time I checked. Still, it doesn' make it that much more complex in this instance, so I've updated it.

File: packages/block-library/src/template-part/edit/name-panel.js
Comment (epiqueras) [reply packages/block-library/src/template-part/edit/name-panel.js:27]: Only if the component is "Pure", when components are shallow and cheap to render like this, making them "Pure" and the memoization overhead can be more expensive than just rerendering.

**Comment by epiqueras** [comment]
>Should the Dropdown / preview panel be wider? It seems very thin, but it does behave well and it doesn't necessarily look bad.

Yeah, I made it wider.

>There are some circumstances where the dropdown has 2 scroll bars. Maybe the height of the preview container should be contained within the dropdown? That is, so the outer scrollbar wouldn't be triggered, the 'search' input would always be at the top of the dropdown, and the 'scroll' would only happen through the preview window itself?

Yes, fixed.

>Also noting nested template parts will trigger the input for their parents. Im not sure if this is good or bad? Probably good in that it improves the visibility that the selected template part is part of another?

Yeah, that's good, it'll also look better when we move it to the toolbar.

>I think we'll need to stick to using the term "Template Part"

Why?

>Let me know when you're ready for a review.

It's ready :smile:

**Comment by MichaelArestad** [comment]
> Why?

I'm worried it could hold up the PR, but maybe not? It is still an experimental feature so we can definitely experiment. Maybe add "template part" as search term so it shows up when searching the inserter

**Comment by epiqueras** [comment]
I think the sooner we switch, the better, and it makes it look a lot better.

File: packages/block-library/src/template-part/edit/name-panel.js
Comment (Addison-Stavlo) [reply packages/block-library/src/template-part/edit/name-panel.js:27]: Ohh right.  It will update when the parent does regardless of prop change anyways.

**Comment by Addison-Stavlo** [comment]
> Maybe add "template part" as search term so it shows up when searching the inserter

🤣  - I was sitting there typing "/template" for the block search multiple times before I realized it was 'Section'

**Comment by MichaelArestad** [comment]
> I think the sooner we switch, the better, and it makes it look a lot better.

Let's do it.

**Review by Addison-Stavlo — CHANGES_REQUESTED** [review]
Im noticing a tab accessibility regression with this implementation.  I am only able to tab through the first couple sections in the preview list before having the dropdown closed by the 5th 'tab':
accessibility-regression

I'm not quite sure what the deal is, it doesn't correspond with the end of the theme group in the preview or anything else.  I can't get past the 4th Section to select any of the further ones.  It does seem to be something with this dropdown as the current preview list tabs through to the end. 🤔

**Comment by MichaelArestad** [comment]
> Also noting nested template parts will trigger the input for their parents. Im not sure if this is good or bad? Probably good in that it improves the visibility that the selected template part is part of another?

@Addison-Stavlo This is a good point I hadn't thought of. It does have me thinking the idea to shift to [a text input in the block toolbar ](https://github.com/WordPress/gutenberg/issues/23213)would be a better approach, but we need more design feedback there first.

**Comment by epiqueras** [comment]
Updated tests :rocket:

**Comment by epiqueras** [comment]
>Im noticing a tab accessibility regression with this implementation.

@Addison-Stavlo For me, it always closes at the last item. But, the same thing happens for patterns, so this is not a regression of this PR.

**Comment by Addison-Stavlo** [comment]
> For me, it always closes at the last item. But, the same thing happens for patterns, so this is not a regression of this PR.

For me it goes all the way through the entire list with the patterns in the block.  With the dropdown it cuts out early. 🤔

**Comment by MichaelArestad** [comment]
This is coming along nicely. I just took this for a spin and so far there are still a few things to do before this is ready to merge:

- [ ]  Make the Section's title edit/save title bar match the Reusable block title bar.

The reusable block title bar looks like this:
image

Right now, this is what I'm seeing for the Section title bar:
image

It looks like it's just an input currently. I would like to see it match the Reusable block UI closely including the Edit/Save button.

- [ ] The section inserter popover has different styling than the block inserter in the master branch.

Here is the latest inserter from the master branch:

image

And here is what I'm seeing for the Section inserter:

image

* The search input styles differ (including a visible label)
* There are extra borders visible around the section list

**Comment by epiqueras** [comment]
>Make the Section's title edit/save title bar match the Reusable block title bar.

How would that work with the "Review Changes" modal? I thought we were moving away from a specific/local save button.

>The section inserter popover has different styling than the block inserter in the master branch.

Yes, these are separate components. I went with what was already being used for template parts. IMO, the current design is not ideal, but it also shouldn't look like the patterns inserter. We have to make it look good for the different heights of template parts and group them by theme and name somehow.

**Comment by MichaelArestad** [comment]
> How would that work with the "Review Changes" modal? I thought we were moving away from a specific/local save button.

I suppose we could do without the edit/save button, but I think they are valuable - particularly for editing the name of the Section. It should still look identical even without the save button which means there's some wrapping container to go around the input.

> Yes, these are separate components. I went with what was already being used for template parts. IMO, the current design is not ideal, but it also shouldn't look like the patterns inserter. We have to make it look good for the different heights of template parts and group them by theme and name somehow.

I understand that they are separate components and that won't look exactly like the current inserter. I do expect it to be pretty close to the mockups. Here was the latest mockup from the proposal:

image

* The search should look identical (including focus/hover/active states) to the block inserter's search.
* The Section previews should have the same border styling as the block patterns and block previews.
* The extra border should be removed.
* Let's keep the Sections single column for now - we can iterate on that later if Sections start getting categories/taxonomies that allow us to optimize space.

**Comment by Addison-Stavlo** [comment]
> How would that work with the "Review Changes" modal?

I would assume if you 'save' on the toolbar, it would no longer appear in the 'review changes'.  Similarly, if you edited and did not hit save, it would then be present in the 'review changes' save flow.

> For me, it always closes at the last item.

So it tabs through all items in the preview list for you?  How many items do you have present in that?  I created a handful of different styled template parts so I had about 8 in mine.  When I first tested it would get through about the first 4 before breakout, after adding another template part it now closes after the first in the list.  There definitely seems to be something funky with the tab flow in the dropdown.

**Comment by epiqueras** [comment]
@MichaelArestad 
>I suppose we could do without the edit/save button, but I think they are valuable - particularly for editing the name of the Section. It should still look identical even without the save button which means there's some wrapping container to go around the input.

I'll update it.

>I understand that they are separate components and that won't look exactly like the current inserter. I do expect it to be pretty close to the mockups. Here was the latest mockup from the proposal:

I was planning to work on that in a follow-up PR since this is already improving on what's already in `master`, and it would have been nice to get it in before the release, but I'll just work on it here now.

@Addison-Stavlo 
>I would assume if you 'save' on the toolbar, it would no longer appear in the 'review changes'. Similarly, if you edited and did not hit save, it would then be present in the 'review changes' save flow.

That feels a bit redundant. @mtias Have you thought about this before?

>So it tabs through all items in the preview list for you? How many items do you have present in that? I created a handful of different styled template parts so I had about 8 in mine. When I first tested it would get through about the first 4 before breakout, after adding another template part it now closes after the very first item in the list. There definitely seems to be something funky with the tab flow in the dropdown.

Yeah, like eight too.

**Comment by MichaelArestad** [comment]
> I was planning to work on that in a follow-up PR since this is already improving on what's already in master, and it would have been nice to get it in before the release, but I'll just work on it here now.

@epiqueras Either way works for me. Let me know how I can help.

**Comment by Addison-Stavlo** [comment]
My only concern is the [tab a11y thing](https://github.com/WordPress/gutenberg/pull/23295#pullrequestreview-434319230).  Everything else looks fantastic!

If others are able to test and confirm they can keyboard navigate through a moderate list of template parts in the dropdown preview, feel free to toss out my change request.  Everything else seems good to me.

**Comment by MichaelArestad** [comment]
I was able to tab through the list of template parts in FF (latest) on Mac (latest). 

In testing, I also noticed the "header" template part, when selected, doesn't display it's name in the text input that appears:

image

**Comment by epiqueras** [comment]
>My only concern is the tab a11y thing. 

This is also broken for the patterns inserter. It's a bigger issue.

**Comment by Addison-Stavlo** [comment]
> I was able to tab through the list of template parts in FF (latest) on Mac (latest).

Nice!  I just cleaned my environment and tried again on FireFox, Chrome, and Safari.  All are the same for me:  I can 'tab' to the first one (header), but the I cannot get to the second one (footer) as it closes the dropdown.

> This is also broken for the patterns inserter. It's a bigger issue.

Ah yeah, in the dropdown but not in the sidebar panel.  It looks like they recently removed the patterns inserter from the dropdown (perhaps for related reasons with a11y?):

`(trying to paste an image of the new inserter on master branch but github isn't happy with uploads atm)`

There is no longer the ability to browse patterns through the dropdown inserter.  Now, there is only a small window where you can search blocks, and if you want patterns it will open the sidebar panel for you to explore them there.  

If we are modeling this style after the dropdown inserter for patterns, and that dropdown inserter has been removed while there is a known 'bigger issue' of a11y problems in place for it... we may need to investigate the reason it was removed and potentially re-think this design?

**Comment by epiqueras** [comment]
@MichaelArestad 

Screen Shot 2020-06-25 at 3 05 22 PM

I updated the styles, and tabbing seems to work better now too.

**Comment by MichaelArestad** [comment]
@epiqueras That looks much better. Thanks!

If you make the input look like this, I think this PR will be good enough to get in:

image

* Add a label to the text input
* Add a container with the same styling as the reusable block input container

**Comment by epiqueras** [comment]
Done

Screen Shot 2020-06-25 at 3 49 27 PM

**Comment by Addison-Stavlo** [comment]
> and tabbing seems to work better now too.

For me it is still closing the dropdown after the first item. 😢 

> This is also broken for the patterns inserter.

Note this has been removed and replaced with the quick inserter which does not seem to have the same issue:

quick-inserter

I wonder if there is anything specific about that implementation that allows the tabbing to work better.

**Comment by epiqueras** [comment]
gif

**Comment by Addison-Stavlo** [comment]
Nice, I am jealous.

Now, if those empty template parts have blocks, nested blocks, nested template parts, etc. does it still work that way?

For me lately (since i cleaned env) I just have 'header' and 'footer' for 2019-blocks and it doesn't get past the header.

**Comment by Addison-Stavlo** [comment]
After testing more, that seems to be the determining factor:

* If it is an 'empty' template part, tab works and moves onto the next one.
* The first time i try to tab past a template part that actually has a block in it, it closes the dropdown.

**Comment by epiqueras** [comment]
@youknowriad Any idea what's going on here?

**Comment by youknowriad** [comment]
@epiqueras I'm not sure what the question is, some context could be helpful. It seems to be about tabbing. The dropdown close  itself if it loses focus, so maybe that's what happening here (focus transferred programmatically elsewhere). For example: when a block gets selected it receives the focus automatically, maybe it's related.

**Comment by epiqueras** [comment]
@youknowriad The popover which shows available template parts with almost identical code to what the patterns inserter uses, closes after tabbing through the first non-empty template part.

**Comment by youknowriad** [comment]
Are you cloning the blocks to preview or not, maybe you're using the same "clientIds" as blocks rendering (which can mean selected too)

**Comment by epiqueras** [comment]
No, the blocks are unique because they are parsed from the template part HTML.

**Comment by epiqueras** [comment]
@Addison-Stavlo Fixed!

Let's ship this!

gif

**Review by Addison-Stavlo — COMMENTED** [review]
This works considerably better.  Do you know what this issue was?  I tried testing / undoing the changes from that last commit but it seems to work better regardless of that (was there a fix elsewhere in the rebase?).

I did notice another issue in a preview containing links (footer from 2019-blocks).  On tabbing, the links steal focus and subsequently close the dropdown again before I can get to the next pattern in the list:

footer-closes


Also - An unrelated note / question regarding how we should handle the namebar for inserting theme supplied template parts.  Currently after inserting the name is blank:

Screen Shot 2020-06-29 at 4 59 56 PM

**Comment by epiqueras** [comment]
>This works considerably better. Do you know what this issue was? I tried testing / undoing the changes from that last commit but it seems to work better regardless of that (was there a fix elsewhere in the rebase?).

The popover was fixed in `master`, I think.

>I did notice another issue in a preview containing links (footer from 2019-blocks). On tabbing, the links steal focus and subsequently close the dropdown again before I can get to the next pattern in the list:

This is a limitation with the block preview that is also present in `master`. It should not block this PR.

>Also - An unrelated note / question regarding how we should handle the namebar for inserting theme supplied template parts. Currently after inserting the name is blank:

It should display correctly now.

**Comment by Addison-Stavlo** [comment]
> This is a limitation with the block preview that is also present in master. It should not block this PR.

Im a bit conflicted on this point.  While it is a limitation with previews, there doesn't seem to be any other implementation of them where this limitation breaks a11y?  So while the limitation with previews itself shouldn't block the PR, the fact that implementing this design forces us to introduce that limitation in an area that creates further concerns with a11y is what makes me feel uneasy.  Does that make sense?

If this is a limitation we can fix, then it seems all will be good.  But if we cannot fix this limitation in previews, then this dropdown preview design itself is further limited and cannot be fully accessible.  Thoughts?

**Comment by epiqueras** [comment]
Patterns with links also break, right?

It shouldn't be difficult to enhance the preview to block tabbing into it.

**Comment by Addison-Stavlo** [comment]
> Patterns with links also break, right?

Im not finding any to test currently... but presumable they would if they existed (in the dropdown / new quick-inserter).

Since we are still under the `experimental` flag with FSE, I suppose it is safe to move forward with this despite the limitation.  But we should look to be aware of this, try to fix the underlying problem, and if all else fails and this is causing issues we can figure out some alternative.

**Review by Addison-Stavlo — APPROVED** [review]
Great job here, thank you for all these updates!

The code looks great and everything tests well from my end, minus the noted underlying limitation with some previews stealing focus (https://github.com/WordPress/gutenberg/pull/23295#issuecomment-651378244).  Since FSE is still 'experimental', it seems ok to move forward with this despite the limitation (the majority of cases seem to work quite well). But we should look to be aware of this, try to fix the underlying problem, and if all else fails and this is causing issues we may need to figure out some alternative to the dropdown down the road.

**Comment by ZebulanStanphill** [comment]
I think it's worth noting that the name "Section" may be confusing for some users who are expecting an HTML `<section>`. It's worth noting that one of the things discussed for the Group block was to add block variations in the future (using different HTML elements), including a Section variation. That would be rather confusing combined with this block _also_ being called "Section". It's less of a problem if the HTML element selector is only provided in the Inspector, but I still question whether "Section" was really the right term to use here.

The naming may also be confusing for people expecting something more like the Group block. In a lot of page builders, "section" refers to a container element, not a global template part.

**Pull request merged**
